### DISCUSSION TITLE
Expand Auto-Numbering Functionality for Catalog Numbers

### DISCUSSION THREAD

**Discussion by grantfitzsimmons** [specify/specify7#5981] [Feature Requests]
**Description:**  
Enhance the auto-numbering system to allow for more complex numbering schemes, including the ability to incorporate prefixes and suffixes while maintaining unique identifiers.

**Acceptance Criteria:**
-  Allow users to define auto-numbering patterns that include prefixes and suffixes.

@realVinayak demonstrated that you can (mis)use the `byyear` numbering system to have an auto-incrementing system based on a generic prefix:

```
<format system="false" name="OSU CE Number Regexp" class="edu.ku.brc.specify.datamodel.CollectingEvent" fieldname="stationFieldNumber" default="true">
<autonumber>edu.ku.brc.af.core.db.AutoNumberGeneric</autonumber>
<field type="regex" size="4" value="[-|0-9]{4}" byyear="true"/>
<field type="constant" size="1" value=":"/>
<field type="numeric" size="4" inc="true"/>
</format>
```

-  Ensure that auto-incrementing can be based on the prefix without breaking existing functionality.
-  Provide clear documentation on how to set up and use the expanded auto-numbering features.

**Comment by melton-jason** [comment]
Correct me if I'm wrong or misunderstanding the requirement, but isn't this functionality already present in Specify? 

https://github.com/user-attachments/assets/4e8098ad-9039-4a16-96bc-b0197bbff063


```xml
  <format system="false" name="PrefixedCatalogNumber" class="edu.ku.brc.specify.datamodel.CollectionObject" fieldname="catalogNumber">
    <field type="regex" size="3" minsize="3" value="[A-Z]{3}" pattern="AAA"/>
    <field type="separator" size="1" value="-"/>
    <field type="numeric" size="6" inc="true"/>
  </format>
```

(There's currently a bug in where the pattern isn't being inserted into the form, but you can still see the behavior of auto-incrementing) 

If the prefix/suffix is static, you can use a constant field instead for the same effect: 

```xml
  <format system="false" name="GeoCatalogNumber" class="edu.ku.brc.specify.datamodel.CollectionObject" fieldname="catalogNumber">
    <field type="constant" size="3" value="MIN"/>
    <field type="separator" size="1" value="-"/>
    <field type="numeric" size="6" inc="true"/>
  </format>
```

The incrementing numeric field will match against other records which have the same prefix/suffix

https://github.com/specify/specify7/blob/6e7bd96b246c3a931b012a212028de75a1ba2664/specifyweb/specify/uiformatters.py#L134-L137

https://github.com/specify/specify7/blob/6e7bd96b246c3a931b012a212028de75a1ba2664/specifyweb/specify/uiformatters.py#L104-L108

https://github.com/specify/specify7/blob/6e7bd96b246c3a931b012a212028de75a1ba2664/specifyweb/specify/uiformatters.py#L189-L268
